[FLINK-9138][filesystem-connectors] Implement time based rollover in BucketingSink

glaksh100 opened the pull request:
## What is the purpose of the change

This pull request enables a time-based rollover of the part file in the BucketingSink. This is particularly applicable when when write throughput is low and helps data become available at a fixed interval, for consumption.

## Brief change log
  - Add a `batchRolloverInterval` field with a setter 
  - Track a `creationTime` for the bucket state
  - Check for `currentProcessingTime` - `creationTime` > `batchRolloverInterval` and roll over if true

## Verifying this change

This change added tests and can be verified as follows:

  - Added a `testRolloverInterval` test method to the `BucketingSinkTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

glaksh100 commented:
@aljoscha @fhueske Can you please take a look?

fhueske commented on flink-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java:
```diff
@@ -87,9 +87,11 @@
- * the part counter is increased and a new part file is created. The batch size defaults to {@code 384MB},
- * this can be configured using {@link #setBatchSize(long)}.
+ * When a part file becomes bigger than the user-specified batch size or when the part file becomes older
+ * than the user-specified roll over interval the current part file is closed,the part counter is increased
```
Add space `closed,the` -> `closed, the`

fhueske commented on flink-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java:
```diff
@@ -87,9 +87,11 @@
- * this can be configured using {@link #setBatchSize(long)}.
+ * When a part file becomes bigger than the user-specified batch size or when the part file becomes older
+ * than the user-specified roll over interval the current part file is closed,the part counter is increased
+ * and a new part file is created. The batch size defaults to {@code 384MB},this can be configured
```
Add space `{@code 384MB},this` -> `{@code 384MB}, this`

fhueske commented on flink-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java:
```diff
@@ -472,6 +480,15 @@ private boolean shouldRoll(BucketState<T> bucketState) throws IOException {
 					subtaskIndex,
 					writePosition,
 					batchSize);
+			} else {
```
update method comment.

fhueske commented on flink-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java:
```diff
@@ -536,6 +553,9 @@ private void openNewPartFile(Path bucketPath, BucketState<T> bucketState) throws
 		}
 
+		// Record the creation time of the bucket
+		bucketState.firstWrittenToTime = processingTimeService.getCurrentProcessingTime();
```
rename parameter to `bucket.creationTime`? At this point, nothing has been written to the file. Actually, the file has not even been created yet.

fhueske replied:
I think the current behavior is also more consistent than starting the timeout when the first record is written.

fhueske commented on flink-connectors/flink-connector-filesystem/src/test/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSinkTest.java:
```diff
@@ -436,6 +463,40 @@ public void testScalingUp() throws Exception {
+		testHarness.setProcessingTime(0L);
+
+		testHarness.processElement(new StreamRecord<>("test1", 1L));
+		checkFs(outDir, 1, 0,  0, 0);
```
rm double blank

fhueske replied:
check other `checkFs()` calls as well

glaksh100 commented:
@fhueske Thank you for reviewing. I have incorporated the changes that include update to the [documentation](https://ci.apache.org/projects/flink/flink-docs-master/dev/connectors/filesystem_sink.html) on the website. 

Travis seems to be failing on a test case that seems unrelated -
```java.lang.AssertionError: This program execution should have failed.
	at org.junit.Assert.fail(Assert.java:88)
	at org.apache.flink.test.misc.SuccessAfterNetworkBuffersFailureITCase.testSuccessfulProgramAfterFailure(SuccessAfterNetworkBuffersFailureITCase.java:75)```

narayaruna commented:
@glaksh100 If you are using BucketingSink with S3, you might be interested in this [PR](https://github.com/apache/flink/pull/4607) as well.

glaksh100 commented:
@fhueske Can you PTAL and merge this PR?

fhueske commented:
Thanks for the update @glaksh100! The changes look good to me. 
What do you think @aljoscha?

kl0u commented on flink-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java:
```diff
@@ -908,6 +929,20 @@ private void handlePendingFilesForPreviousCheckpoints(Map<Long, List<String>> pe
+	 */
+	public BucketingSink<T> setBatchRolloverInterval(long batchRolloverInterval) {
+		this.batchRolloverInterval = batchRolloverInterval;
+		return this;
```
Please add checks for invalid configs like negative values.

glaksh100 replied:
Added a check for `batchRolloverInterval` to be a positive non-zero value.

aljoscha commented on flink-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java:
```diff
@@ -473,6 +482,15 @@ private boolean shouldRoll(BucketState<T> bucketState) throws IOException {
 					writePosition,
 					batchSize);
+			} else {
+				long currentProcessingTime = processingTimeService.getCurrentProcessingTime();
```
You should pass in the current processing time to avoid calling this somewhat costly method twice.

glaksh100 replied:
Updated method signature for `shouldRoll` to include the `currentProcessingTime`

glaksh100 commented:
Thanks for reviewing @fhueske  @aljoscha and @kl0u ! I have addressed the latest review comments. Can you PTAL (again) ?

fhueske commented:
Hi @glaksh100, 

I just noticed that the bucket closing check is only done when a record is written. Hence, inactive buckets might not get closed in time if a larger inactive bucket interval is configured. In some sense, the new feature is an extended version of the inactive bucket closing feature.

How should we handle that case?

1. throw an exception during configuration, i.e., when `setInactiveBucketThreshold()` and `setBatchRolloverInterval()` are called.
2. configure the inactive bucket interval to be at least the rollover interval in case it is configured larger and continue. We should also make sure that the check interval is configured appropriately.

I'm leaning towards the first approach. It would make the misconfiguration obvious to the user and fail the program before it is submitted.

What do you think?

Best, Fabian

glaksh100 commented:
@fhueske  I gave it some thought and your suggestion makes sense to me. I have extended `checkForInactiveBuckets` to include the rollover check. I have also updated Javadocs in a few places:
- Added a note in the top-level Javadocs to update functionality of `checkForInactiveBuckets()`
- Updated JavaDocs for both `setBatchRolloverInterval()` and `setInactiveBucketThreshold()`
- Updated JavaDoc for `checkForInactiveBuckets()`

Let me know if the updates make sense and thank you for reviewing!

fhueske commented:
Hi @glaksh100, thanks for the update!

fhueske commented:
Merging
